oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/optimizer/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,151 @@
+    ) = responseObserver.consumeThen({
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.requestStart(request)
+        }
+    }) { response ->
+        if (response.acknowledged) apiService.start()
+    }
+
+    override fun stopOptimization(
+            request: StopOptimizationCommandDTO,
+            responseObserver: StreamObserver<StopOptimizationResponseDTO>
+    ) = responseObserver.consumeThen({
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.requestStop(request)
+        }
+    }) { response ->
+        if (response.responseCase == StopOptimizationResponseDTO.ResponseCase.RUNID) apiService.stop()
+    }
+
+    override fun updateNode(
+            request: NodeStatusCommandOrResponseDTO,
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.updateNode(request)
+        }
+    }
+
+    override fun sendMessage(
+            request: MessageCommandDTO,
+            responseObserver: StreamObserver<MessageResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.sendMessage(request)
+        }
+    }
+
+    override fun unregisterRequest(
+            request: RequestUnRegistrationRequestDTO,
+            responseObserver: StreamObserver<UnRegistrationResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.unregister(request)
+        }
+    }
+
+    override fun offerSimulationResult(
+            request: SimulationResponseDTO,
+            responseObserver: StreamObserver<SimulationResultConfirmDTO>
+    ) = responseObserver.consumeAsync {
+        checkThenRunAsync(modelService.simulations.hasName(request.name)) {
+            apiService.offerResult(request)
+        }
+    }
+
+    override fun offerErrorResult(
+            request: ErrorResponseDTO,
+            responseObserver: StreamObserver<ErrorConfirmDTO>
+    ) = responseObserver.consumeAsync {
+        checkThenRunAsync(modelService.simulations.hasName(request.name)) {
+            apiService.offerError(request)
+        }
+    }
+
+    override fun offerSimulationConfig(
+            request: NodeStatusCommandOrResponseDTO,
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consumeAsync {
+        checkThenRunAsync(modelService.simulations.hasName(request.name)) {
+            apiService.offerConfig(request)
+        }
+    }
+
+    private fun <V> checkThenRun(hasPermission: Boolean, action: () -> V): V {
+        if (hasPermission) {
+            return action()
+        } else {
+            throw StatusRuntimeException(Status.PERMISSION_DENIED)
```

Comment:
oh it throws. It doesnt go unhappy path, it goes error. I see...

Reply:
maybe change it from `checkThenRun` to `assertThenRun` or `requireThenRun`.

Reply:
but honestly vincent, why not simply express

```kt
checkThenRun(service.someCondition()){ apiService.someResponseFactory() }
```

as

```kt
require(service.someCondition())
apiService.someResponseFactory()
```

?

Reply:
ah, I guess that is also an option, but I think i was trying to go for the `StatusRuntimeException(Status.PERMISSION_DENIED)` when looking into being more specific but I guess it is also can be described as an IllegalArgumentException

Reply:
also, offense programming does make the code a lot different, I did this a bit backward and there is a lot of defences check down the line where they are now becoming useless. I will add more require to replace those.